rotkehlchen/chain/evm/decoding/compound/v3/balances.py

```diff
@@ -0,0 +1,142 @@
+        """
+        Fetch unique borrow events from the userDB. Since a user can increase or decrease the same
+        liability, we remove the duplicates to reduced the amount of queries. Returns a dict of
+        compound token's address -> set of addresses which borrowed the underlying token, and a
+        dict of compound token's address -> its underlying token.
+        """
+        unique_borrows: dict['EvmToken', set['ChecksumEvmAddress']] = defaultdict(set)
+        underlying_tokens: dict['ChecksumEvmAddress', 'EvmToken'] = {}
+        for address, events in self.addresses_with_activity(
+            event_types={(HistoryEventType.RECEIVE, HistoryEventSubType.GENERATE_DEBT)},
+        ).items():
+            for event in events:
+                if event.address is None:
+                    continue
+                try:
+                    compound_token = EvmToken(evm_address_to_identifier(
+                        address=event.address,
+                        chain_id=self.evm_inquirer.chain_id,
+                        token_type=EvmTokenKind.ERC20,
+                    ))
+                    if len(compound_token.underlying_tokens) == 0:
+                        log.error(f'No underlying token found for {compound_token!s}.')
+                        continue
+
+                    underlying_token = EvmToken(evm_address_to_identifier(
+                        address=compound_token.underlying_tokens[0].address,
+                        chain_id=self.evm_inquirer.chain_id,
+                        token_type=EvmTokenKind.ERC20,
+                    ))
+                except (UnknownAsset, WrongAssetType) as e:
+                    log.error(
+                        "Failed to resolve compound v3 borrow event's token and/or its "
+                        f'underlying token in {event.tx_hash.hex()} due to {e!s}',
+                    )
+                    continue
+
+                unique_borrows[compound_token].add(address)
+                underlying_tokens[event.address] = underlying_token
+
+        return unique_borrows, underlying_tokens
+
+    def query_balances(self) -> 'BalancesSheetType':
+        """
+        Query liabilities for Compound v3 open positions and return them. The assets are handled
+        by the wrapped compound v3 tokens, so those are not queried here.
+
+        It calls the Compound token contracts to get the borrow balances of the addresses and
+        tokens whose borrow history events are found in the userDB.
+        """
+        balances: BalancesSheetType = defaultdict(BalanceSheet)
+        unique_borrows, underlying_token = self._extract_unique_borrowed_tokens()
+        if len(unique_borrows) == 0:
+            return balances
+
+        # prepare the arguments for the contract calls and the encoded call
+        calls: list[tuple['ChecksumEvmAddress', str]] = []
+        calls_arguments: list[tuple['ChecksumEvmAddress', 'EvmToken']] = []
```

Comment:
It may be beneficial to have this as a named tuple.

Code will become much more readable if that was the case and all the comments about what `[calls_arguments[idx][0]` etc. are would just be `.attribute` thus understandable without the need for comments.

I also normally avoid named tuples unless something is repeated many times